dap/DapUtils.cc

```diff
@@ -77,4 +80,147 @@ log_request_and_memory_size(/*const*/ DMR &dmr)
+        stringstream msg;
+        msg << endl;
+        msg << "ERROR: Unable to convert a DAP4 DMR for this dataset to a DAP2 DDS object. " << endl;
+        msg << "This dataset contains variables and/or attributes whose data types are not compatible " << endl;
+        msg << "with the DAP2 data model." << endl;
+        msg << endl;
+        msg << "There are " << inventory.size() << " incompatible variables and/or attributes referenced " << endl;
+        msg  << "in your request." << endl;
+        msg << "Incompatible variables: " << endl;
+        msg << endl;
+        for(const auto &entry: inventory){
+            msg << "    " << entry << endl;
+        }
+        throw BESSyntaxUserError(msg.str(), __FILE__, __LINE__);
+    }
+}
+
+void throw_for_dap4_typed_attrs(DAS *das)
+{
+    vector<string> inventory;
+    if(has_dap4_typed_attributes("", das->container(), inventory)){
+        stringstream msg;
+        msg << endl;
+        msg << "ERROR: Unable to convert a DAP4 DMR for this dataset to a DAP2 DAS object. " << endl;
+        msg << "This dataset contains attributes whose data types are not compatible " << endl;
+        msg << "with the DAP2 data model." << endl;
+        msg << endl;
+        msg << "There are " << inventory.size() << " incompatible attributes referenced " << endl;
+        msg  << "in your request." << endl;
+        msg << "Incompatible attributes: " << endl;
+        msg << endl;
+        for(const auto &entry: inventory){
+            msg << "    " << entry << endl;
+        }
+        throw BESSyntaxUserError(msg.str(), __FILE__, __LINE__);
+    }
+}
+
+
+bool has_dap4_typed_attributes(const std::string &path, AttrTable *atable, std::vector<std::string> &inventory)
+{
+    if(!atable)
+        return false;
+
+    bool has_d4_attr = false;
+    for(auto aitr=atable->attr_begin();aitr!=atable->attr_end();aitr++){
+        bool ima_d4_attr = false;
+        string attr_fqn = path + "@" + (*aitr)->name;
+        switch ((*aitr)->type) {
+            case Attr_int8:
+            case Attr_int64:
+            case Attr_uint64:
+            case Attr_enum:
+            case Attr_opaque:
+            {
+                ima_d4_attr = true;
+                break;
+            }
+            case Attr_container:
+            {
+                ima_d4_attr |= has_dap4_typed_attributes(attr_fqn, (*aitr)->attributes, inventory);
+                break;
+            }
+            default:
+                //noop;
+                break;
+        }
+        if(ima_d4_attr){
+            inventory.emplace_back(AttrType_to_String((*aitr)->type) + " " + attr_fqn);
+        }
+        has_d4_attr |= ima_d4_attr;
+    }
+    return has_d4_attr;
+}
+
+/**
+ * @brief convenience function for the response limit test.
+ * The DDS stores the response size limit in Bytes even though the context
+ * param uses KB. The DMR uses KB throughout.
+ * @param dds
+ */
+void throw_if_dap2_response_too_big(DDS *dds)
```

Comment:
This all looks good, but if the file and line info was passed into these 'throw functions' the original source would not be lost.